### ISSUE TITLE
Investigation of multiple object detection services

### ISSUE THREAD

**Issue by vgibilmanno** [artiso-solutions/CoVoX#42]
# Azure Computer Vision

**Language**: C#
**Training required?**: No
**Input**: Image as stream to API Endpoint
**Returns**: JSON containing info about detected objects (with descriptions), bounding boxes and overall image description
**Good for live video feed?**: No since doing a frame per frame detection would consume too many API endpoint calls.

Works only for really basic and day to day used objects like bottle, cups etc. If I want to detect a cactus it doesn't work.

![image](https://user-images.githubusercontent.com/19838802/108978652-1d0eb680-768a-11eb-9bfe-9640497de58a.png)

# Azure Custom Vision

**Language**: C#
**Training required?**: Yes
You have to upload images and label all objects of every image. This can be done programmatically by providing a JSON with bounding box regions and a label or with an online image tagger provided by MSFT.
Training a cactus worked well and the detection was kinda ok. The training dataset wasn't distinct enough to provide satisfactory results though.

**Input**: Image as stream to API Endpoint
**Returns**: JSON containing info about detected objects (with descriptions), bounding boxes and overall image description
**Good for live video feed?**: No since doing a frame per frame detection would consume too many API endpoint calls.

![image](https://user-images.githubusercontent.com/19838802/108979423-ebe2b600-768a-11eb-87b1-2bb1cd3e629f.png)

# Offline Model based on YOLOv3 CPU (ImageAI)
https://github.com/OlafenwaMoses/ImageAI

**Language**: Python
**Training required?**: Optional
YOLOv3 has only 80 classes which it was trained with which represent most day to day objects (bottle, cups, cars etc). For more specific objects like a cactus you have to individually re-train the model.

**Input**: Image as stream to library
**Returns**: Object containing info about detected objects (with descriptions) and bounding boxes
**Good for live video feed?**: Yes though very slow. With fastest approach I was able to get 2FPS

![image](https://user-images.githubusercontent.com/19838802/108979737-3d8b4080-768b-11eb-8902-0aabcdea565c.png)

# Offline Model based on MobileNet SSD CPU

**Language**: Python
**Training required?**: Optional

**Input**: Image as stream to library
**Returns**: Object containing info about detected objects (with descriptions) and bounding boxes
**Good for live video feed?**: Very fast. Probably best model for live video prediction

The results though are very unreliable and work only for really basic things but training this model on a custom dataset could prove to be good.

Video: https://streamable.com/3qh0mf

![image](https://user-images.githubusercontent.com/19838802/108989848-59481400-7696-11eb-9349-2806edfb9fb5.png)

# Offline Model based on YOLOv5 CPU (PyTorch)
https://github.com/OlafenwaMoses/ImageAI

**Language**: Python
**Training required?**: Optional
YOLOv5 like YOLOv3 has only 80 classes which it was trained with which represent most day to day objects (bottle, cups, cars etc). For more specific objects like a cactus you have to individually re-train the model.

**Input**: Image as stream to library
**Returns**: Object containing info about detected objects (with descriptions) and bounding boxes
**Good for live video feed?**: Yes much faster than YOLOv3. I get multiple FPS. So it's slower than MobilNet SSD but more precise

![image](https://user-images.githubusercontent.com/19838802/109020482-49dac200-76ba-11eb-9302-80008bafafec.png)

# How to label images for offline models
To label images for offline models there is Roboflow which does a really good job.
https://app.roboflow.com/

After labeling all images you can export them in the format of your desired model.

![image](https://user-images.githubusercontent.com/19838802/109273756-e57f4600-7812-11eb-8b34-c0e4359f47d7.png)
![image](https://user-images.githubusercontent.com/19838802/109273823-ff208d80-7812-11eb-8c52-beff05bfd907.png)
